docs/: Add instructions on contributing to documentation (#171)
unikraft/docs: content/en/docs/contributing/docs.md

## Documentation

As with the other parts of Unikraft, we welcome contributions to the [documentation repository](https://github.com/unikraft/docs).
Documentation is written using [GitHub Markdown](https://github.github.com/gfm/) and is built using [Hugo](https://gohugo.io/commands/hugo_server/).

We follow the [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow) for contributions.
Submit corrections and improvements to documentation as pull requests (PRs) that we will then review and approve in the documentation.
You can add entire sections if you think a topic is missing or update existing ones with missing information.

You can also open up an issue to signal a problem or a missing part in the documentation that you request someone to fix.

> We use English as a documentation language.

Reviewer: ```suggestion
We use English as the documentation language.
```

    |
    |-- community/              # Community related content
    |   |-- contacts.md         # Community contacts
    |   |-- events.md           # Information related to recurring and upcoming events
    |   |-- github.md           # Unikraft GitHub community layout
    |   |-- hackathons/         # Hackathon sessions, challenges and solutions
    |   |-- _index.md           # Community first page
    |   |-- people.md           # List of people in the community
    |   |-- publications.md     # List of Unikraft-related publications (theses, peer-reviewed publications, etc.)
    |   `-- talks.md            # List of Unikraft-related talks and presentations
    |
    |-- docs/                   # Unikraft documentation, split by categories into subdirectories
    |
    |-- faq/                    # List of frequently asked questions
    |
    `-- releases/               # Unikraft releases summaries
```

### Testing Changes

Before you submit your changes for review, please make sure you have tested them thoroughly.
You can build and deploy the site locally, following [the instructions below](/docs/contributing/#building-the-website).

> You can also use the [`super-linter`](https://github.com/github/super-linter), following the [instructions below](docs/contributing/docs/#using-the-linter).

Reviewer: ```suggestion
You can also validate your changes for common issues using [`super-linter`](https://github.com/github/super-linter), following the [instructions below](docs/contributing/docs/#using-the-linter).
```

Please also do a spell check on your changes.
Most text editors should be able to do that by themselves, [here's how you do it in `Vim`](https://thoughtbot.com/blog/vim-spell-checking).

#### Building the Website

You can build the website inside a Docker environment or natively.
Building inside a Docker environment is recommended, since you will work in the same environment as the deployed website.

#### Building the Website Inside a Docker Environment

You can build the site inside a Docker environment.
For this, you will need to install Docker-CE.
You can do that by following the instructions [here](https://github.com/docker/docker-install), or by running:

```console
$ curl -fsSL https://get.docker.com/ | sh
```

> {{% alert title="Note" color="warning" %}}

Reviewer: ```suggestion
{{% alert theme="info" %}}
```

Building the site inside a Docker environment is highly recommended, since you will work in the same environment as the deployed site, so there will be no errors regarding packages version or filesystem layout.
{{% /alert %}}

To deploy the site, run (in the directory of the repository clone):

```console
$ make container
$ make devenv # This will take you inside the container

# Inside the container, run the following commands:
# npm install
# make serve
```

You may need admin privileges to run the first 2 commands.

#### Building the Website Natively

You can also build the site natively.
For a native build, you need to install `hugo`.
You can do it by following the instructions [here](https://gohugo.io/getting-started/installing/).

Deploying the website requires [`npm`](https://www.npmjs.com/) packages (listed in `package.json`).
First, install `npm` for your distribution;
for a Debian/Ubuntu distribution, run:

```console
$ sudo apt install npm
```

Then install required packages:

```console
$ npm install # Run from the documentation clone directory
```

> {{% alert title="Note" color="warning" %}}

Reviewer: ```suggestion
{{% alert theme="warning" %}}
```

>  his may lead to errors due to different `hugo` versions or other environment problems.

Reviewer: ```suggestion
This may lead to errors due to different `hugo` versions or other environment problems.
```

> The problems can be avoided by building the website inside a Docker environment.

Reviewer: ```suggestion
These problems can be avoided by building the website inside a Docker environment.
```

If you choose to build the site natively, make sure you install and use [`Hugo v0.98.0`](https://github.com/gohugoio/hugo/releases/tag/v0.98.0) or later.
{{% /alert %}}

Then, to build and deploy the site, run:

```console
$ hugo serve
```

After building and deploying the site, `hugo` provides instructions on accessing and using it:

```text
Web Server is available at http://localhost:1313/ (bind address 127.0.0.1)
```

Access the `http://localhost:1313/` URL in the browser to view the site.

> #### Using the linter

Reviewer: ```suggestion
#### Using the Linter
```

To automate a big part of the reviewing process, we use the [`super-linter`](https://github.com/github/super-linter) on the `docs` repository.
It runs as a GitHub action on every time you do a push, checking all the changed files.

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter.png"
   title="Checking the super linter output"
   caption="If you open a pull request, you will be able to see the linter output under the checks section."
   position="center"
>}}

> If you click on `details`, you can see the full linter output.

Reviewer: ```suggestion
If you click on `Details`, you can see the full linter output.
```

The most relevant part is under the `Lint Code Base` section, where you can find the full error messages and the line that generated the error.

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter-output.png"
   title="Checking the super linter output"
   caption="Full linter output. Error was caused by line 79 in file `content/en/blog/2022-12-03-random-device-support-in-unikraft.md`"
   position="center"
>}}

Since the linter checks the entire file, you may find yourself with a lot of errors from lines that you did not modify.
Please make sure that the errors are not caused by the lines that you added or modified.
If you have the time to add a commit to your pull request, that fixes all the linting errors within the modified files, we will be forever grateful :pray:

> ##### Running the Linter Locally

Reviewer: ```suggestion
#### Running the Linter Locally
```

You can also run the linter locally, before creating the pull request.
[`super-linter`](https://github.com/github/super-linter) is actually a collection of linters;
for a local run, you need to install and then run the required linters.

For the Unikraft documentation repository, used and useful linters are:

- [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli#readme)
- [`textlint`](https://textlint.github.io/)
- [`jscpd`](https://github.com/kucherenko/jscpd)

To run the `super-linter` locally, follow the instructions from [here](https://github.com/github/super-linter/blob/main/docs/run-linter-locally.md).
Note that this will require quite a lot of storage, since it will pull the `super-linter` Docker image.

You will need to install Docker, following the [official documentation](https://docs.docker.com/install/).
After that, pull the latest container by running:

```console
$ docker pull github/super-linter:latest
```

> To run the `super-linter` for one file, use:

Reviewer: ```suggestion
To run the `super-linter` on a given directory, use a command similar to:
```

2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   The Super-Linter source code can be found at:
2022-12-29 07:50:55 [INFO]    - https://github.com/github/super-linter
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   --------------------------------------------
[...]
```

A sample run, while in the root directory of this documentation repository, is:

```console
$ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -v $(pwd)/content/en/docs/contributing/:/tmp/lint/ github/super-linter
```

You can also run each linter separately, since it will require less storage space.
To run the linters, you will need to install the [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli), [`textlint`](https://textlint.github.io/docs/getting-started.html) and [`jscpd`](https://github.com/kucherenko/jscpd#installation) tools:

```console
$ npm install markdownlint-cli
$ npm install textlint
$ npm install jscpd
```

> Note that these linters require a more recent version of `node` than the default package will install.

Reviewer: ```suggestion
Note that it's likely that these linters require a more recent version of `node` than the default package installation.
```

Install the most recent one using [`nvm`](https://github.com/nvm-sh/nvm).
First, install `nvm` using [the instructions here](https://github.com/nvm-sh/nvm).
Then install the most recent version of `node` and follow the instructions:

```console
$ nvm install node

> # ... follow instructions from https://github.com/nvm-sh/nvm#usage

Reviewer: ```suggestion
# Follow instructions shown by the above command.
```

You can also open up an issue to signal a problem or a missing part in the documentation that you request someone to fix.

We use English as a documentation language.
Please proofread your text before submitting a pull request.

Write each sentence on a new line.
This way, changing one sentence only affects one line in the source code.

Prefer using the 1st person plural.
Use phrases like "we run the command / app", "we look at the source code", "we provide support".

### Content Layout

> The content is structured as such:

Reviewer 2: ```suggestion
The content has the following structure:
```

    |   |-- _index.md           # Community first page
    |   |-- people.md           # List of people in the community
    |   |-- publications.md     # List of Unikraft-related publications (theses, peer-reviewed publications, etc.)
    |   `-- talks.md            # List of Unikraft-related talks and presentations
    |
    |-- docs/                   # Unikraft documentation, split by categories into subdirectories
    |
    |-- faq/                    # List of frequently asked questions
    |
    `-- releases/               # Unikraft releases summaries
```

### Testing Changes

Before you submit your changes for review, please make sure you have tested it thoroughly.
You can build and deploy the site locally, following [the instructions below](/contributing/docs/#building).
You can also use the [Super Linter](https://github.com/github/super-linter), following the [instructions below](docs/contributing/docs/#using-the-linter).
Please also do a spell check on your changes.
Most text editors should be able to do that by themselves, [here's how you do it in `Vim`](https://thoughtbot.com/blog/vim-spell-checking).

> #### Building the website

Reviewer: ```suggestion
#### Building the Website
```

You can build the website inside a Docker environment or natively.
Building inside a Docker environment is recommended, since you will work in the same environment as the deployed website.

> #### Building the website inside a Docker environment

Reviewer: ```suggestion
#### Building the Website Inside a Docker Environment
```

You can build the site inside a Docker environment.
For this, you will need to install Docker-CE.
You can do that by following the instructions [here](https://github.com/docker/docker-install), or by running:

```console
$ curl -fsSL https://get.docker.com/ | sh
```

**Note: Building the site inside a Docker environment is highly recommended, since you will work in the same environment as the deployed site, so there will be no errors regarding packages version or filesystem layout.**

> To then deploy the site, run:

Reviewer 2: ```suggestion
To deploy the site, run:
```

```console
$ make container
$ make devenv # This will take you inside the container

> # To be run inside the docker env

Reviewer 2: ```suggestion
# Inside the container,  run the following commands:
```

$ npm install
$ make serve
```

You may need admin privileges to run the first 2 commands.

> #### Building the website natively

Reviewer: ```suggestion
#### Building the Website Natively
```

You can also build the site natively.
For a native build, you need to install `hugo`.

> You can do it by following the instructions [here](https://gohugo.io/getting-started/installing/).

Reviewer 2: ```suggestion
You can do it by following the instructions from [gohugo.io/getting-started/installing](https://gohugo.io/getting-started/installing/).
```

Then, to build and deploy the site, run:

```console
$ hugo serve
```

**Note: This may lead to errors due to different `hugo` versions or other environment problems.
The problems can be avoided by building the webside inside a Docker environment.
If you choose to build the site natively, make sure you use [`Hugo v0.98.0`](https://github.com/gohugoio/hugo/releases/tag/v0.98.0).**

After building and deploying the site, `hugo` provides instructions on accessing and using it:

```text

> Web Server is available at http://localhost:1313/ (bind address 127.0.0.1)

Reviewer 2: Could we provide a link here to that address, to be one-click away?

Author: I would not provide a link to a localhost address, people might get confused if it does not work.
The link will be available from the command output in their terminal.

#### Using the linter

To automate a big part of the reviewing process, we use the [super linter](https://github.com/github/super-linter) on the `docs` repository.
It runs as a GitHub action on every time you do a push, checking all the changed files.

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter.png"
   title="Checking the super linter output"
   caption="If you open a pull request, you will be able to see the linter output under the checks section."
   position="center"
>}}

If you click on `details`, you can see the full linter output.
The most relevant part is under the `Lint Code Base` section, where you can find the full error messages and the line that generated the error.

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter-output.png"
   title="Checking the super linter output"
   caption="Full linter output. Error was caused by line 79 in file `content/en/blog/2022-12-03-random-device-support-in-unikraft.md`"
   position="center"
>}}

Since the linter checks the entire file, you may find yourself with a lot of errors from lines that you did not modify.

> If you have the time to add a commit to your pull request that fixes all the linting errors within the modified files, we will be forever grateful :pray:.

Reviewer: ```suggestion
If you have the time to add a commit to your pull request that fixes all the linting errors within the modified files, we will be forever grateful :pray:
```

> If you have the time to add a commit to your pull request that fixes all the linting errors within the modified files, we will be forever grateful :pray:.

Reviewer: ```suggestion
If you have the time to add a commit to your pull request that fixes all the linting errors within the modified files, we will be forever grateful :pray:
```

If not, please make sure that the errors are not caused by the lines that you added or modified.

> **You can also run the linter locally.**

Reviewer: ```suggestion
**You can also run the linter locally.
Note that there is a set of linters run via super-linter.
For this documentation repository, used and useful linters are [`markdown-cli`](TODO), [`textlint`](TODO), [`jscpd`](TODO).
Instructions on installing and using them are below.**

Also add instructions on running super-linter locally, but note that it requires a large storage space (for its container).
```

For this, you will need to install the [`markdown-cli` tool](https://github.com/igorshubovych/markdownlint-cli).

```console
$ npm install -g markdownlint-cli
```

To run the linter, from inside the directory you want to check, use:

```console
$ markdownlint --config <path-to-your-docs-repo-clone>/.github/workflows/config/config.json --rules <path-to-your-docs-repo-clone>/.github/workflows/rules/rules.js *.md
```

> ### Reviewing changes

Reviewer: ```suggestion
### Reviewing Changes
```

If you are assigned as a reviewer to a pull request, please follow the [testing steps](docs/contributing/docs/#testing-changes) and make sure everything looks good.

> The review may go through a back-and-forth between the authors and the reviewer before the reviewer marks the PR as approved.

Reviewer: ```suggestion
The review may go through a set of back-and-forth steps between the author and the reviewer before the reviewer marks the PR as approved.
```

> The review may go through a back-and-forth between the authors and the reviewer before the reviewer marks the PR as approved.

Reviewer: ```suggestion
The review may go through a set of back-and-forth steps between the author and the reviewer before the reviewer marks the PR as approved.
```

As a reviewer, when you consider the process complete you must add a signature, in the same way that the author has.

> This is done by writing adding a `Reviewed-by` Git trailer like so:

Reviewer: ```suggestion
This is done by writing adding a `Reviewed-by` line, as part of your approval message in the GitHub review interface:
```

```text
Reviewed-by: Your Name <your@email.com>

> ```

Reviewer 2: I think there should be somebody automatically assigned when a PR is made. @StefanJum @razvand do you know if we can do this the same way we have on other Unikraft's repos?

Reviewer: Yes, we could do that via the [`governance` scripts](https://github.com/unikraft/governance/), but I wouldn't prioritize this.

Author: I agree this is nice to have, but not a priority, since there are not that many prs on the docs repo being opened at the moment.

To deploy the site, run:

```console
$ make container
$ make devenv # This will take you inside the container

# Inside the container, run the following commands:
$ npm install
$ make serve
```

You may need admin privileges to run the first 2 commands.

#### Building the Website Natively

You can also build the site natively.
For a native build, you need to install `hugo`.
You can do it by following the instructions [here](https://gohugo.io/getting-started/installing/).

Then, to build and deploy the site, run:

```console
$ hugo serve
```

**Note: This may lead to errors due to different `hugo` versions or other environment problems.
The problems can be avoided by building the webside inside a Docker environment.
If you choose to build the site natively, make sure you use [`Hugo v0.98.0`](https://github.com/gohugoio/hugo/releases/tag/v0.98.0).**

After building and deploying the site, `hugo` provides instructions on accessing and using it:

```text
Web Server is available at http://localhost:1313/ (bind address 127.0.0.1)
```

Access the `http://localhost:1313/` URL in the browser to view the site.

#### Using the linter

> To automate a big part of the reviewing process, we use the [super linter](https://github.com/github/super-linter) on the `docs` repository.

Reviewer: ```suggestion
To automate a big part of the reviewing process, we use the [`super-linter`](https://github.com/github/super-linter) on the `docs` repository.
```

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter.png"
   title="Checking the super linter output"
   caption="If you open a pull request, you will be able to see the linter output under the checks section."
   position="center"
>}}

If you click on `details`, you can see the full linter output.
The most relevant part is under the `Lint Code Base` section, where you can find the full error messages and the line that generated the error.

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter-output.png"
   title="Checking the super linter output"
   caption="Full linter output. Error was caused by line 79 in file `content/en/blog/2022-12-03-random-device-support-in-unikraft.md`"
   position="center"
>}}

Since the linter checks the entire file, you may find yourself with a lot of errors from lines that you did not modify.
If you have the time to add a commit to your pull request that fixes all the linting errors within the modified files, we will be forever grateful :pray:
If not, please make sure that the errors are not caused by the lines that you added or modified.

**You can also run the linter locally.

> Note that there is a set of linters run via super-linter.

Reviewer: ```suggestion
Note that there is a set of linters run via `super-linter`.
```

```console
$ docker pull github/super-linter:latest
```

To run the `super-linter` for one file, use:

```console
$ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -v /path/to/local/codebase/file:/tmp/lint/file github/super-linter
--------------------------------------------------------------------------------
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   --- GitHub Actions Multi Language Linter ----
2022-12-29 07:50:55 [INFO]    - Image Creation Date:[2022-12-26T16:46:33Z]
2022-12-29 07:50:55 [INFO]    - Image Revision:[154522b380449499473f75d46f3796e23ebc1d2e]
2022-12-29 07:50:55 [INFO]    - Image Version:[154522b380449499473f75d46f3796e23ebc1d2e]
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   The Super-Linter source code can be found at:
2022-12-29 07:50:55 [INFO]    - https://github.com/github/super-linter
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   --------------------------------------------
[...]
```

> You can also run each liner separately, since it will require less storage space.

Reviewer: ```suggestion
You can also run each linter separately, since it will require less storage space.
```

> To run the linters, you will need to install the [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli), [`textlint`](https://textlint.github.io/docs/getting-started.html) and [`jscpd`](https://github.com/kucherenko/jscpd#installation) tools.

Reviewer: ```suggestion
To run the linters, you will need to install the [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli), [`textlint`](https://textlint.github.io/docs/getting-started.html) and [`jscpd`](https://github.com/kucherenko/jscpd#installation) tools:
```

2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   --------------------------------------------
[...]
```

You can also run each linter separately, since it will require less storage space.
To run the linters, you will need to install the [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli), [`textlint`](https://textlint.github.io/docs/getting-started.html) and [`jscpd`](https://github.com/kucherenko/jscpd#installation) tools:

```console
$ npm install -g markdownlint-cli
$ npm install -g textlint
$ npm install -g jscpd
```

The `textlinter` will require installation of rules.
You can find a list of rules [here](https://github.com/textlint/textlint/wiki/Collection-of-textlint-rule).

```console
$ npm install -g textlint-rule-no-todo
```

To run the linters, from inside the directory you want to check, use:

```console
$ markdownlint --config <path-to-your-docs-repo-clone>/.github/workflows/config/config.json --rules <path-to-your-docs-repo-clone>/.github/workflows/rules/rules.js *.md
$ npx textlint --rule <rule-name> <file>
$ npx jscpd <path-to-source-code>

> ```

Reviewer: Add sample runs, concrete commands and their output, to let readers know what they should fill the `<path-to-source-code>` argument with. And to have a feel of what the output should look like.

### Reviewing Changes

> If you are assigned as a reviewer to a pull request, please follow the [testing steps](docs/contributing/docs/#testing-changes) and make sure everything looks good.

Reviewer: ```suggestion
If you are assigned as a reviewer to a pull request (PR), please follow the [testing steps](docs/contributing/docs/#testing-changes) and make sure everything looks good.
```

The review may go through a set of back-and-forth steps between the author and the reviewer before the reviewer marks the PR as approved.

> As a reviewer, when you consider the process complete you must add a signature, in the same way that the author has.

Reviewer: ```suggestion
As a reviewer, when you consider the process complete, you must add a signature, in the same way that the author has.
```

You can also open up an issue to signal a problem or a missing part in the documentation that you request someone to fix.

We use English as a documentation language.
Please proofread your text before submitting a pull request.

Write each sentence on a new line.
This way, changing one sentence only affects one line in the source code.

Prefer using the 1st person plural.
Use phrases like "we run the command / app", "we look at the source code", "we provide support".

### Content Layout

The content has the following structure:

```text
content/
`-- en
    |-- blog/                   # Blog posts, <date>-<blogpost title>.md
    |
    |-- community/              # Community related content
    |   |-- contacts.md         # Community contacts

>     |   |-- events.md           # Informations related to reccuring and upcoming events

Reviewer: ```suggestion
    |   |-- events.md           # Information related to recurring and upcoming events
```

    |   |-- github.md           # Unikraft GitHub community layout

>     |   |-- hackathons/         # Hackathon sessions, challenges and sollutions

Reviewer: ```suggestion
    |   |-- hackathons/         # Hackathon sessions, challenges and solutions
```

    |   |-- _index.md           # Community first page
    |   |-- people.md           # List of people in the community
    |   |-- publications.md     # List of Unikraft-related publications (theses, peer-reviewed publications, etc.)
    |   `-- talks.md            # List of Unikraft-related talks and presentations
    |
    |-- docs/                   # Unikraft documentation, split by categories into subdirectories
    |
    |-- faq/                    # List of frequently asked questions
    |
    `-- releases/               # Unikraft releases summaries
```

### Testing Changes

> Before you submit your changes for review, please make sure you have tested it thoroughly.

Reviewer: ```suggestion
Before you submit your changes for review, please make sure you have tested them thoroughly.
```

#### Building the Website Inside a Docker Environment

You can build the site inside a Docker environment.
For this, you will need to install Docker-CE.
You can do that by following the instructions [here](https://github.com/docker/docker-install), or by running:

```console
$ curl -fsSL https://get.docker.com/ | sh
```

**Note: Building the site inside a Docker environment is highly recommended, since you will work in the same environment as the deployed site, so there will be no errors regarding packages version or filesystem layout.**

To deploy the site, run:

```console
$ make container
$ make devenv # This will take you inside the container

# Inside the container, run the following commands:
$ npm install
$ make serve
```

You may need admin privileges to run the first 2 commands.

#### Building the Website Natively

You can also build the site natively.
For a native build, you need to install `hugo`.
You can do it by following the instructions [here](https://gohugo.io/getting-started/installing/).

Then, to build and deploy the site, run:

```console
$ hugo serve
```

**Note: This may lead to errors due to different `hugo` versions or other environment problems.

> The problems can be avoided by building the webside inside a Docker environment.

Reviewer: ```suggestion
The problems can be avoided by building the website inside a Docker environment.
```

If you have the time to add a commit to your pull request that fixes all the linting errors within the modified files, we will be forever grateful :pray:
If not, please make sure that the errors are not caused by the lines that you added or modified.

**You can also run the linter locally.
Note that there is a set of linters run via `super-linter`.**

For the Unikraft documentation repository, used and useful linters are:

- [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli#readme)
- [`textlint`](https://textlint.github.io/)
- [`jscpd`](https://github.com/kucherenko/jscpd)

To run the `super-linter` locally, follow the instructions from [here](https://github.com/github/super-linter/blob/main/docs/run-linter-locally.md).
Note that this will require quite a lot of storage, since it will pull the `super-linter` Docker image.

You will need to install Docker, following the [official documentation](https://docs.docker.com/install/).
After that, pull the latest container by running:

```console
$ docker pull github/super-linter:latest
```

To run the `super-linter` for one file, use:

```console

> $ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -v /path/to/local/codebase/file:/tmp/lint/file github/super-linter

Reviewer: ```suggestion
$ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -v /path/to/directory/to/check/:/tmp/lint/ github/super-linter
```

--------------------------------------------------------------------------------
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   --- GitHub Actions Multi Language Linter ----
2022-12-29 07:50:55 [INFO]    - Image Creation Date:[2022-12-26T16:46:33Z]
2022-12-29 07:50:55 [INFO]    - Image Revision:[154522b380449499473f75d46f3796e23ebc1d2e]
2022-12-29 07:50:55 [INFO]    - Image Version:[154522b380449499473f75d46f3796e23ebc1d2e]
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   The Super-Linter source code can be found at:
2022-12-29 07:50:55 [INFO]    - https://github.com/github/super-linter
2022-12-29 07:50:55 [INFO]   ---------------------------------------------
2022-12-29 07:50:55 [INFO]   --------------------------------------------
[...]
```

> 

Reviewer: ```suggestion
A sample run, while in the root directory of this documentation repository, is:

"""
$ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -v $(pwd)/content/en/docs/contributing/:/tmp/lint/ github/super-linter
"""
```

You can also run each linter separately, since it will require less storage space.
To run the linters, you will need to install the [`markdown-cli`](https://github.com/igorshubovych/markdownlint-cli), [`textlint`](https://textlint.github.io/docs/getting-started.html) and [`jscpd`](https://github.com/kucherenko/jscpd#installation) tools:

```console
$ npm install -g markdownlint-cli
$ npm install -g textlint
$ npm install -g jscpd
```

The `textlinter` will require installation of rules.
You can find a list of rules [here](https://github.com/textlint/textlint/wiki/Collection-of-textlint-rule).

```console
$ npm install -g textlint-rule-no-todo
```

To run the linters, from inside the directory you want to check, use:

```console
$ markdownlint --config <path-to-your-docs-repo-clone>/.github/workflows/config/config.json --rules <path-to-your-docs-repo-clone>/.github/workflows/rules/rules.js *.md
$ npx textlint --config <path-to-textlint-config-file> --rule <rule-name> <file>
$ npx jscpd <path-to-source-code>
```

> For example:

Reviewer: Awesome example :+1:

```console
$ git clone https://github.com/unikraft/docs
$ cd docs/content/en/docs/contributing
$ markdownlint --config ../../../../.github/workflows/config/config.json --rules ../../../../.github/workflows/rules/rules.js *.md
_index.md:14:187 MD033/no-inline-html Inline HTML [Element: a]
blog-case-studies.md:10 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
code-of-conduct.md:69 MD104/one line per sentence one line (and only one line) per sentence [Expected one sentence per line. Multiple end of sentence punctuation signs found on one line!]
code-of-conduct.md:82:1 MD007/ul-indent Unordered list indentation [Expected: 0; Actual: 1]
code-of-conduct.md:85:1 MD007/ul-indent Unordered list indentation [Expected: 0; Actual: 1]
[...]

> $ textlint --config ../../../../.github/workflows/.textlintrc code-of-conduct.md

Reviewer: ```suggestion
$ npx textlint --config .github/workflows/.textlintrc --rule terminology content/en/docs/contributing/code-of-conduct.md
```

> /media/stefan/projects/unikraft/docs/content/en/docs/contributing/code-of-conduct.md

Reviewer: ```suggestion
[...]/docs/content/en/docs/contributing/code-of-conduct.md
```

  61:66  ✓ error  Incorrect usage of the term: “e-mail”, use “email” instead  terminology

✖ 1 problem (1 error, 0 warnings)
✓ 1 fixable problem.
Try to run: $ textlint --fix [file]

> $ jscpd code-of-conduct.md

Reviewer: ```suggestion
$ npx jscpd content/en/docs/contributing/code-of-conduct.md
```

Prefer using the 1st person plural.
Use phrases like "we run the command / app", "we look at the source code", "we provide support".

### Content layout

The content is structured as such:

```tree
content/
`-- en
    |-- blog                    # Blog posts, <date>-<blogpost title>.md
    |
    |-- community               # Community related content.
    |   |-- contacts.md         # Community contacts.
    |   |-- events.md           # Informations related to reccuring and upcoming events.
    |   |-- github.md           # Unikraft GitHub community layout.
    |   |-- hackathons/         # Hackathon sessions, challenges and sollutions.
    |   |-- _index.md           # Community first page.
    |   |-- people.md           # List of people in the community.
    |   |-- publications.md     # List of Unikraft-related publications (theses, peer-reviewed publications, etc.).
    |   `-- talks.md            # List of Unikraft-related talks and presentations.
    |
    |-- docs                    # Unikraft documentation, split by categories into subdirectories.
    |
    |-- faq                     # List of frequently asked questions.
    |

>     `-- releases                # Unikraft releases summaries.

Reviewer: Add a slash (`/`) at the end of directory names. Remove dot (`.`) at the end of comments (it should only be used for sentences).

```

### Testing changes

Before you submit your changes for review, please make sure you have tested it thoroughly.
You can build and deploy the site locally, following [the instructions below](/contributing/docs/#building).
You can also use the [Super Linter](https://github.com/github/super-linter), following the [instructions below](docs/contributing/docs/#using-the-linter).
Please also do a spell check on your changes.
Most text editors should be able to do that by themselves, [here's how you do it in `Vim`](https://thoughtbot.com/blog/vim-spell-checking).

#### Building the site

> 

Reviewer: Add a section for building inside a container.

You can build the site inside a Docker environment.
For this, you will need to install Docker-CE.
You can do that by following the instructions [here](https://github.com/docker/docker-install), or by running:

```console
$ curl -fsSL https://get.docker.com/ | sh
```

**Note: Building the site inside a Docker environment is highly recommended, since you will work in the same environment as the deployed site, so there will be no errors regarding packages version or filesystem layout.**

To then deploy the site, run:

```console
$ make container
$ make devenv

# To be run inside the docker env
$ npm install
$ make serve
```

You may need admin privileges to run the first 2 commands.

> 

Reviewer: Add a section for building natively.

You can also build the site natively.
For a native build, you need to install `hugo`.
You can do it by following the instructions [here](https://gohugo.io/getting-started/installing/).

Then, to build and deploy the site, run:

```console
$ hugo serve
```

**Note: This may lead to errors due to different `hugo` versions or other environment problems.
The problems can be avoided by building the webside inside a Docker environment.
If you choose to build the site natively, make sure you use [`Hugo v0.98.0`](https://github.com/gohugoio/hugo/releases/tag/v0.98.0).**

After building and deploying the site, `hugo` provides instructions on accessing and using it:

```text
Web Server is available at http://localhost:1313/ (bind address 127.0.0.1)
```

Access the `http://localhost:1313/` URL in the browser to view the site.

#### Using the linter

To automate a big part of the reviewing process, we use the [super linter](https://github.com/github/super-linter) on the `docs` repository.

> It runs as a GitHub action on every new commit you make, checking all the changed files.

Reviewer: "every time you do a push"

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter.png"
   title="Checking the super linter output"
   caption="If you open a pull request, you will be able to see the linter output under the checks section."
   position="center"
>}}

If you click on `details`, you can see the full linter output.
The most relevant part is under the `Lint Code Base` section, where you can find the full error messages and the line that generated the error.

{{< img
   class="w-auto mx-auto"
   src="/assets/imgs/github-super-linter-output.png"
   title="Checking the super linter output"
   caption="Full linter output. Error was caused by line 79 in file `content/en/blog/2022-12-03-random-device-support-in-unikraft.md`"
   position="center"
>}}

Since the linter checks the entire file, you may find yourself with a lot of errors from lines that you did not modify.

> If you have the time to add a commit to your pull request that fixes all the linting errors within the modified files, we will be forever grateful.

Reviewer: Add a :pray: to show a bit of fun to the latter wording.

If not, please make sure that the errors are not caused by the lines that you added or modified.

**You can also run the linter locally.**
For this, you will need to install the [`markdown-cli` tool](https://github.com/igorshubovych/markdownlint-cli).

```console
$ npm install -g markdownlint-cli
```

To run the linter, from inside the directory you want to check, use:

```console
$ markdownlint --config <path-to-your-docs-repo-clone>/.github/workflows/config/config.json --rules <path-to-your-docs-repo-clone>/.github/workflows/rules/rules.js *.md
```

### Reviewing changes

If you are assigned as a reviewer to a pull request, please follow the [testing steps](docs/contributing/docs/#testing-changes) and make sure everything looks good.

The review may go through a back-and-forth between the authors and the reviewer before the reviewer marks the PR as approved.

> As a reviewer, when you consider the process complete you must add an equivalent `sign off` in the same way that the author has.

Reviewer: Do not say "an equivalent "sign off"" as it will confuse the reader. Just say: you must add a signature.
